### PR TITLE
11032 - List Key Commands tool for Editor

### PR DESCRIPTION
**Pull request by Cattlesquat**
Lists the key commands. Lets you sort/filter them. 

ALSO... gave Chatter a right+click context menu for copy (you could already copy with Ctrl+C so just a bit of polish)

### CODE REVIEW

File: vassal-app/src/main/java/VASSAL/build/module/Chatter.java [hunk 1]
```diff
@@ -686,9 +725,539 @@ public static void main(String[] args) {
...
+        new ParserDelegator().parse(reader, parserCallback, true);
+      }
+      catch (IOException e) {
+        e.printStackTrace();
+      }
+      String result = "";
+      for (final String s : list) {
+        result += s;
+      }
+      return result;
```
Comment (uckelman) [inline vassal-app/src/main/java/VASSAL/build/module/Chatter.java]: You could `return String.join("", list);` here.

File: vassal-app/src/main/java/VASSAL/build/module/Chatter.java [hunk 2]
```diff
@@ -686,9 +725,539 @@ public static void main(String[] args) {
...
+  /**
+   * @deprecated use {@link GlobalOptions#getPlayerId()}
+   */
   @Deprecated(since = "2020-08-06", forRemoval = true)
   public String getHandle() {
     return GlobalOptions.getInstance().getPlayerId();
   }
+
+
+  static class ChatTransferHandler extends TransferHandler {
```
Comment (uckelman) [inline vassal-app/src/main/java/VASSAL/build/module/Chatter.java]: `private static class`?

File: vassal-app/src/main/java/VASSAL/build/module/Chatter.java [hunk 3]
```diff
@@ -686,9 +725,539 @@ public static void main(String[] args) {
...
+      }
+    }
+
+    @Override
+    public int getSourceActions(JComponent c) {
+      return COPY;
+    }
+  }
+
+  static class ChatTransferable implements Transferable {
```
Comment (uckelman) [inline vassal-app/src/main/java/VASSAL/build/module/Chatter.java]: `private static class`?

File: vassal-app/src/main/java/VASSAL/build/module/Chatter.java [hunk 4]
```diff
@@ -686,9 +725,539 @@ public static void main(String[] args) {
...
+      }
+    }
+
+    public DataFlavor[] getTransferDataFlavors() {
+      return supportedFlavors;
+    }
+
+    public boolean isDataFlavorSupported(DataFlavor flavor) {
+      for (final DataFlavor supportedFlavor : supportedFlavors) {
+        if (supportedFlavor == flavor) {
```
Comment (uckelman) [inline vassal-app/src/main/java/VASSAL/build/module/Chatter.java]: Should this be `supportedFlavor.equals(flavor)` or are you really intending to compare by identity?

File: vassal-app/src/main/java/VASSAL/build/module/Chatter.java [hunk 5]
```diff
@@ -686,9 +725,539 @@ public static void main(String[] args) {
...
+      String[] mimeTypes = {
+        "text/html", "text/plain"
+      };
+      Class<?>[] textClasses = {
+        Reader.class, String.class, CharBuffer.class, char[].class
+      };
+      Class<?>[] byteClasses = {
+        InputStream.class, ByteBuffer.class, byte[].class
+      };
+      String[] charsets = {
```
Comment (uckelman) [inline vassal-app/src/main/java/VASSAL/build/module/Chatter.java]: Can these arrays be final?

File: vassal-app/src/main/java/VASSAL/build/module/Chatter.java [hunk 6]
```diff
@@ -686,9 +725,539 @@ public static void main(String[] args) {
...
+    @Override
+    public int getSourceActions(JComponent component) {
+      return COPY_OR_MOVE;
+    }
+
+    @Override
+    public void exportToClipboard(JComponent component,
+                                  Clipboard clipboard,
+                                  int action) {
+      JTextPane pane = (JTextPane) component;
```
Comment (uckelman) [inline vassal-app/src/main/java/VASSAL/build/module/Chatter.java]: final?

File: vassal-app/src/main/java/VASSAL/build/module/Chatter.java [hunk 7]
```diff
@@ -686,9 +725,539 @@ public static void main(String[] args) {
...
+        appendTextContent(element.getElement(i), textStart, textEnd,
+          content);
+      }
+    }
+  }
+
+
+
+
+  class StyleTransferHandler extends TransferHandler {
```
Comment (uckelman) [inline vassal-app/src/main/java/VASSAL/build/module/Chatter.java]: private?

File: vassal-app/src/main/java/VASSAL/build/module/Chatter.java [hunk 8]
```diff
@@ -686,9 +725,539 @@ public static void main(String[] args) {
...
+    }
+  }
+
+  public class HTMLCopier
+    extends TransferHandler {
+    private static final long serialVersionUID = 1;
+
+    private final Collection<DataFlavor> flavors;
+
+    HTMLCopier() {
```
Comment (uckelman) [inline vassal-app/src/main/java/VASSAL/build/module/Chatter.java]: `public`?

File: vassal-app/src/main/java/VASSAL/build/module/Chatter.java [hunk 9]
```diff
@@ -686,9 +725,539 @@ public static void main(String[] args) {
...
+        return styledStringFlavor.equals(flavor);
+      }
+
+      public String toString() {
+        return "StyledStringTransferable: " + styledString;
+      }
+    }
+  }
+
+  class StyledString {
```
Comment (uckelman) [inline vassal-app/src/main/java/VASSAL/build/module/Chatter.java]: private?

File: vassal-app/src/main/java/VASSAL/build/module/Chatter.java [hunk 10]
```diff
@@ -686,9 +725,539 @@ public static void main(String[] args) {
...
+        sb.append(" for " + locs.get(j));
+        if (j < attrs.size()-1)
+          sb.append("\n");
+      }
+      return sb.toString();
+    }
+  }
+
+
+  class Location {
```
Comment (uckelman) [inline vassal-app/src/main/java/VASSAL/build/module/Chatter.java]: private?

**Review by uckelman — CHANGES_REQUESTED** [review]
There are a bunch of inner classes that need their visibility checked.

**Comment by Cattlesquat** [comment]
Wait, how is any of that scratch stuff still in there -- I show chatter as clean in that branch -- have you pulled latest?

**Comment by uckelman** [comment]
> Wait, how is any of that scratch stuff still in there -- I show chatter as clean in that branch -- have you pulled latest?

I was commenting on a particular commit, I think. Let me look again.

File: vassal-app/src/main/java/VASSAL/build/AbstractConfigurable.java [hunk 11]
```diff
@@ -259,4 +261,27 @@ public Configurer getConfigurer() {
...
+   */
+  public String getTypeName() {
+    try {
+      return (String) this.getClass().getMethod("getConfigureTypeName").invoke(null);
+    }
+    catch (NoSuchMethodException e) {
+      // Ignore. This is normal, since some classes won't have this method.
+    }
+    catch (IllegalAccessException | ExceptionInInitializerError
+      | NullPointerException | InvocationTargetException | IllegalArgumentException e) {
```
Comment (uckelman) [inline vassal-app/src/main/java/VASSAL/build/AbstractConfigurable.java:281]: Is an NPE actually possible here?

File: vassal-app/src/main/java/VASSAL/build/AbstractConfigurable.java [hunk 12]
```diff
@@ -259,4 +261,27 @@ public Configurer getConfigurer() {
...
+      ErrorDialog.bug(e);
+    }
+
+    return this.getClass().getName().substring(this.getClass().getName().lastIndexOf(".") + 1);
```
Comment (uckelman) [inline vassal-app/src/main/java/VASSAL/build/AbstractConfigurable.java:285]: I think you can use `this.getClass().getSimpleName()` here.

**Comment by Cattlesquat** [comment]
Although I experimented with some bespoke transfer handlers, none of them gave me what I *really* wanted (transferring formatted text to native apps) and so I reverted to the default Copy action.

File: vassal-app/src/main/resources/VASSAL/i18n/VASSAL.properties [hunk 13]
```diff
@@ -58,6 +58,7 @@ General.properties=Properties
 General.test=Test
 General.exit=Exit
 General.run=Run
+General.copy=Copy
```
Comment (uckelman) [inline vassal-app/src/main/resources/VASSAL/i18n/VASSAL.properties:61]: Huh, we didn't have Copy already.

**Review by uckelman — CHANGES_REQUESTED** [review]

**Comment by uckelman** [comment]
Kids these days, experimenting with their bespoke transfer handlers...

File: vassal-app/src/main/java/VASSAL/build/AbstractConfigurable.java (hunk 11 above)
Comment (Cattlesquat) [reply vassal-app/src/main/java/VASSAL/build/AbstractConfigurable.java:281]: I'm not sure -- I cut and paste this directly out of ConfigureTree, didn't write it myself.

File: vassal-app/src/main/java/VASSAL/build/AbstractConfigurable.java (hunk 12 above)
Comment (Cattlesquat) [reply vassal-app/src/main/java/VASSAL/build/AbstractConfigurable.java]: Done

**Review by uckelman — APPROVED** [review]

**Pull request merged**